test/ERC20CosmosCoinAmountConversionTest.t.sol

```diff
@@ -0,0 +1,436 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.25 <0.9.0;
+
+// solhint-disable custom-errors,max-line-length
+import { Test } from "forge-std/Test.sol";
+import { ERC20CosmosCoinAmountConversion } from "../src/utils/ERC20CosmosCoinAmountConversion.sol";
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+// Discuss - Do we want to move mock contract to a mock folder?
+// Mock ERC20 token without the decimals function overridden
+contract MockERC20 is ERC20 {
+    constructor() ERC20("BaseToken", "BTK") {
+        _mint(msg.sender, 100_000_000_000_000_000_000); // Mint some tokens for testing
+    }
+}
+
+// Discuss - Do we want to move mock contract to a mock folder?
+// Mock ERC20 token with the decimals function overridden
+contract MockERC20Metadata is ERC20 {
+    uint8 private _decimals;
+
+    constructor(uint8 decimals_) ERC20("MetadataToken", "MTK") {
+        _decimals = decimals_;
+        _mint(msg.sender, 100_000_000_000_000_000_000); // Mint some tokens for testing
+    }
+
+    function decimals() public view override returns (uint8) {
+        return _decimals;
+    }
+}
+
+contract ERC20CosmosCoinAmountConversionTest is Test {
+    // Instance of the MockERC20 contract
+    MockERC20 private mockERC20;
+    MockERC20Metadata private mockERC20Metadata;
+
+    function setUp() public {
+        // Deploy the mock ERC20 tokens
+        mockERC20 = new MockERC20();
+        mockERC20Metadata = new MockERC20Metadata(18);
+    }
+
+    // Tests for MockERC20 - Standard Tokens implementation without decimals --> decimals = 18
+
+    /**
+     * Problem Breakdown
+     * ERC20 to Cosmos Coin Conversion:
+     * We are converting an ERC20 token amount to a Cosmos coin amount.
+     * - ERC20 standard use 18 decimals.
+     * - Cosmos coins use 6 decimals.
+     * Example Conversion:
+     * - Input: 1000000000000000001 (1 ERC20 token + 1 extra smallest unit, i.e., 1 Wei).
+     * - Expected Converted Amount: 1000000 (1 ERC20 token should convert to 1,000,000 Cosmos coins).
+     * - Expected Remainder: 1 (which is the remaining smallest unit that doesn't fit into the Cosmos coin format).
+     * The Remainder is what we espect to return to the user
+     * Detailed Explanation:
+     * - Given 1000000000000000001 as input:
+     * Conversion Calculation:
+     * - 1 ERC20 token = 1000000000000000000 (in Wei).
+     * - Converting to Cosmos coin units involves dividing by 10^12
+     * (since ERC20 has 18 decimals and Cosmos has 6 decimals, the difference is 18 - 6 = 12).
+     * Mathematical Conversion:
+     * - Converted Amount: 1000000000000000001 / 10^12 = 1000000 (Cosmos coins).
+     * - Remainder: 1000000000000000001 % 10^12 = 1 (Wei remaining).
+     */
+    function testConvertMockERC20TokenAmountToCosmosCoin_1() public view {
```

Comment:
It might change a bit if you decide upon using the fuzzing framework as many of these tests could be written as a single test, but just a general comment: I personally prefer the test functions to be more clearly named what scenario/sub-scenario they are testing, rather than 1, 2, 3 etc.

Reply:
Definitely agree. Reasons for this naming are [here](https://github.com/srdtrk/solidity-ibc-eureka/pull/19/files#r1705624096)

Reply:
Yes, please rename the tests. It'd be good to refactor these to use table test pattern if possible. You can see me do this [here](https://github.com/srdtrk/solidity-ibc-eureka/blob/main/test/IbcIdentifiers.t.sol).

Reply:
Does the fuzz tests with fixtures do the same @srdtrk?

https://book.getfoundry.sh/forge/fuzz-testing#fuzz-test-fixtures